test/e2e/test/elasticsearch/checks_k8s.go

```diff
@@ -254,3 +227,154 @@ func CheckESPassword(b Builder, k *test.K8sClient) test.Step {
+			// But **not** be annotated with the hash of a different ES spec, meaning
+			// it probably still matches the spec of the pre-mutation builder (rolling upgrade not over).
+			//
+			// Important: this does not catch rolling upgrades due to a keystore change, where the Builder hash
+			// would stay the same.
+			expectedHash := nodeSetHash(b.Elasticsearch, nodeSet)
+			if p.Annotations[BuilderHashAnnotation] != "" && p.Annotations[BuilderHashAnnotation] != expectedHash {
+				return fmt.Errorf("pod %s was not upgraded (yet?) to match the expected Elasticsearch specification", p.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func checkStatefulSetsReplicas(b Builder, k *test.K8sClient) error {
+	// build names and replicas count of expected StatefulSets
+	expected := make(map[string]int32, len(b.Elasticsearch.Spec.NodeSets)) // map[StatefulSetName]Replicas
+	for _, nodeSet := range b.Elasticsearch.Spec.NodeSets {
+		expected[esname.StatefulSet(b.Elasticsearch.Name, nodeSet.Name)] = nodeSet.Count
+	}
+	statefulSets, err := k.GetESStatefulSets(b.Elasticsearch.Namespace, b.Elasticsearch.Name)
+	if err != nil {
+		return err
+	}
+	// compare with actual StatefulSets
+	actual := make(map[string]int32, len(statefulSets)) // map[StatefulSetName]Replicas
+	for _, statefulSet := range statefulSets {
+		actual[statefulSet.Name] = *statefulSet.Spec.Replicas // should not be nil
+	}
+	if !reflect.DeepEqual(expected, actual) {
+		return fmt.Errorf("invalid StatefulSets: expected %v, got %v", expected, actual)
+	}
+	return nil
+}
+
+func AnnotatePodsWithBuilderHash(b Builder, k *test.K8sClient) []test.Step {
+	return []test.Step{
+		{
+			Name: "Annotate Pods with a hash of their Builder spec",
+			Test: func(t *testing.T) {
+				annotateWithRetries(b, k, t)
+			},
+		},
+		// make sure this is propagated to the local cache so next test steps can expect annotated pods
+		{
+			Name: "Wait for annotated Pods to appear in the cache",
+			Test: test.Eventually(func() error {
+				pods, err := sset.GetActualPodsForCluster(k.Client, b.Elasticsearch)
+				if err != nil {
+					return err
+				}
+				for _, p := range pods {
+					if p.Annotations[BuilderHashAnnotation] == "" {
+						return fmt.Errorf("pod %s is not annotated with %s yet", p.Name, BuilderHashAnnotation)
+					}
+				}
+				return nil
+			}),
+		},
+	}
+}
+
+func annotateWithRetries(b Builder, k *test.K8sClient, t *testing.T) {
+	es := b.Elasticsearch
+	for _, nodeSet := range b.Elasticsearch.Spec.NodeSets {
+		pods, err := sset.GetActualPodsForStatefulSet(k.Client, types.NamespacedName{
+			Namespace: es.Namespace,
+			Name:      esname.StatefulSet(es.Name, nodeSet.Name),
+		})
+		require.NoError(t, err)
+		for i := range pods {
+			pods[i].Annotations[BuilderHashAnnotation] = nodeSetHash(es, nodeSet)
+			err := k.Client.Update(&pods[i])
+			if err != nil && apierrors.IsConflict(err) {
+				// pod was updated concurrently, retry
+				annotateWithRetries(b, k, t)
```

Comment:
I have mixed feelings about this:) can we just loop over k.Client.Update n times instead?

Reply:
We need to loop over Get + Update (not only Update). So basically loop over the entire function. I can wrap it in our `test.Eventually` helper though, which should simplify it a bit and add a retry timeout.